File: lim/다익스트라/Dijkstra - 닷 투더 닷.py
'''
c의 최소값을 고정하고 다익스트라를 돌린다. 실행후 실제 최소값이 c가 아닌 경우는 갱신에 적용 안 한다
c최소값을 간선들의 c 값들로 변경시켜가며 위 과정을 진행하고 각각의 계산값들중 최소값을 구해 출력
'''

import heapq
import math
import sys
from collections import defaultdict


def calculate_value(dist, min_c):
    global x
    return dist + x / min_c


def get_min(n, c_limit, graph):
    min_values = [sys.maxsize] * (n+1)
    min_values[1] = 0
    min_cs = [sys.maxsize] * (n+1)
    min_cs[1] = sys.maxsize
    hq = [(0, 0, sys.maxsize, 1)]  # (계산값, 거리, c, 노드)
    while len(hq):
        value, dist, min_c, node = heapq.heappop(hq)
        if value > min_values[node]:
            continue
        for n_node, l, c in graph[node]:
            if c < c_limit:
                continue
            n_min_c = min(min_c, c)
            n_dist = dist + l
            n_value = calculate_value(n_dist, c_limit)
            if min_values[n_node] > n_value:
                min_values[n_node] = n_value
                min_cs[n_node] = n_min_c
                heapq.heappush(hq, (n_value, n_dist, n_min_c, n_node))
    if c_limit != min_cs[n]:
        return None
    return min_values[n]


def sol(n, m, cs, graph):
    min_value = sys.maxsize
    for c_limit in cs:
        value = get_min(n, c_limit, graph)
        if value:
            min_value = min(min_value, value)
    print(math.floor(min_value))


n, m, x = map(int, sys.stdin.readline().split())
graph = defaultdict(list)
cs = []
for _ in range(m):
    s, e, l, c = map(int, sys.stdin.readline().split())
    graph[s].append((e, l, c))
    graph[e].append((s, l, c))
    cs.append(c)

sol(n, m, cs, graph)

File: lim/다익스트라/Dijkstra_닷_투더_닷.java
package lim.다익스트라;

import java.util.*;
import java.io.*;

public class Dijkstra_닷_투더_닷 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int n, m, x;
    static ArrayList<ArrayList<Cost>> graph = new ArrayList<>();
    static HashSet<Integer> cValues = new HashSet<>();

    static class Cost {
        int dest, l, c;

        Cost(int dest, int l, int c) {
            this.dest = dest;
            this.l = l;
            this.c = c;
        }
    }

    static class Item implements Comparable<Item> {
        int node, dist;

        public Item(int node, int dist) {
            this.node = node;
            this.dist = dist;
        }

        @Override
        public int compareTo(Item item) {
            return this.dist - item.dist;
        }
    }

    static void dijkstra(int sNode, int lowerBoundC, int[] dists) {
        PriorityQueue<Item> queue = new PriorityQueue<>();
        queue.offer(new Item(sNode, 0));
        dists[sNode] = 0;

        while (!queue.isEmpty()) {
            Item item = queue.poll();
            if (item.dist > dists[item.node]) {
                continue;
            }

            if (graph.get(item.node) == null) {
                continue;
            }
            for (Cost cost : graph.get(item.node)) {
                int nNode = cost.dest;
                int nDist = item.dist + cost.l;
                if (dists[nNode] <= nDist || cost.c < lowerBoundC) {
                    continue;
                }
                dists[nNode] = nDist;
                queue.offer(new Item(nNode, dists[nNode]));
            }
        }
    }

    static int getMinTime(int lowerBoundC) {
        int[] dists = new int[n + 1];
        for (int i = 0; i < dists.length; i++) {
            dists[i] = Integer.MAX_VALUE;
        }
        dijkstra(1, lowerBoundC, dists);
        if (dists[n] == Integer.MAX_VALUE) {
            return Integer.MAX_VALUE;
        }
        return dists[n] + x / lowerBoundC;
    }

    static void sol() throws IOException {
        int minTime = Integer.MAX_VALUE;
        for (int cValue : cValues) {
            minTime = Math.min(minTime, getMinTime(cValue));
        }
        bw.write(minTime + "");
    }

    public static void main(String[] args) throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        x = Integer.parseInt(st.nextToken());

        for (int i = 0; i < n + 1; i++) {
            graph.add(new ArrayList<>());
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int s = Integer.parseInt(st.nextToken());
            int e = Integer.parseInt(st.nextToken());
            int l = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());

            graph.get(s).add(new Cost(e, l, c));
            graph.get(e).add(new Cost(s, l, c));
            cValues.add(c);
        }
        sol();
        br.close();
        bw.flush();
        bw.close();
    }
}